### ISSUE TITLE
Poor performance of MicrometerCollector.collect()

### ISSUE THREAD

**Issue by Hartigan** [micrometer-metrics/micrometer#3993]
Poor performance of `MicrometerCollector.collect()` which uses `DefaultLongTaskTimer`. `DefaultLongTaskTimer` contains a lot of calls of `ConcurrentLinkedDeque.size()` with `O(n)` complexity during histogram creation.

Micrometer version: 1.10.3

Might be worth saving the size of deque or full snapshot at the beginning of the method `DefaultLongTaskTimer.takeSnapshot()` execution. Also method may contain data races between histogram creation and updates of `activeTasks`.

**Comment by jonatan-ivanov** [comment]
Thank you for the issue!

Do you have benchmarks that shows the bottleneck or is this a theoretical issue?
I would assume that you will have issues because of the huge amount of parallel tasks before you would see performance degradations because of `size()`. This assumption might be incorrect, some benchmarks can demonstrate the correctness of it.

>Also method may contain data races between histogram creation and updates of activeTasks.

Could you please elaborate this? Which method and where can race condition arise?

**Comment by Hartigan** [comment]
True, I have service with huge amount of parallel requests and it's expected load (not issue). And `ConcurrentLinkedDeque.size()` consumes 8% of flamegraph -- actual issue. 
<img width="713" alt="Screenshot 2023-07-25 at 10 19 15" src="https://github.com/micrometer-metrics/micrometer/assets/3034112/ede82a58-0a06-4e0f-90e3-a563f66aff12">

I calculate the algorithmic complexity of this class:
`DefaultLongTaskTimer` store task in `ConcurrentLinkedDeque` and provide methods `start()` with `O(1) for append task into deque`, `SampleImpl.stop()` -- `O(n) for removing random element from deque` and `takeSnapshot()` -- `O(n^2) for iterating over deque with size() calls on each iteration`.

> Could you please elaborate this? Which method and where can race condition arise?

It's theoretical, but you use `ConcurrentLinkedDeque`. This implies that timers can be added and removed while the histogram is being calculated(calls of `start()/SampleImpl.stop()` during call of `takeSnapshot()`).

**Comment by Hartigan** [comment]
I created draft pull request https://github.com/micrometer-metrics/micrometer/pull/4001

**Comment by Hartigan** [comment]
If you may sacrifice `DefaultLongTaskTimer#max()` performance, I can replace `Deque` by `HashMap` and improve performance of `stop` operation. For cases with huge amount of active tasks I see 3-5% of profile in `ConcurrentLinkedDeque#remove(Object)`, because tasks located not at the beginning of deque. Example: 100 active long requests and timer start/stop new short requests. `start()` -- just append new timer for short request, and `stop()` must iterate 100 long task for delete short task each time.

**Comment by fdulger** [comment]
Is this planned to be taken up anytime soon? We're also experiencing similar behavior when we let our application run for a long time. We're using spring boot and micrometer with default configuration and we scrape the metrics with a prometheus server once in a minute. We notice that the CPU usage of the system is going up continuously. If we increase the frequency of scraping, the CPU usage goes up faster. First, the micrometer endpoint starts to respond slower and eventually the entire application gets effected.

**Comment by Sipaha** [comment]
@fdulger We had the same issue in our system. The problem started after we updated Spring, which now registers its own DefaultMeterObservationHandler. We already had one handler registered, so this caused a conflict.
Each handler registers samples in the onStart method, but only the last one properly stops them in onStop. The others stay in a 'started' state forever, leading to a memory leak and increased CPU usage, just like what you're seeing.

**Comment by jonatan-ivanov** [comment]
Fyi, since `1.13.0`, you can disable `LongTaskTimer` creation: https://github.com/micrometer-metrics/micrometer/issues/4604. If you are using Spring Boot, there is a property for that:
[`actuator.management.observations.long-task-timer.enabled`](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#application-properties.actuator.management.observations.long-task-timer.enabled).

Also, I would really look into why some of your `Observation`s are never stopped (it's likely that there is a problem with your instrumentation). Not stopping `Observation`s or `LongTaskTimer`s are in fact memory leaks and it's a separate problem than this issue (it's likely a problem with the instrumentation not Micrometer).
